### PR TITLE
ramda: R.has() shall be usable as a type guard

### PR DESCRIPTION
**Pull request by kulla**
#### Template

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://ramdajs.com/docs/#has

#### About this PR

This PR changes the signature of [`R.has()`](https://ramdajs.com/docs/#has) so that it can be used as a type guard. An example:

```ts
const data: unknown = { ... }

if (R.has("name", foo)) {
  console.log(data.name) // name can now be accessed
}
```

In the changes of `types/ramda/test/has-tests.ts` you see other examples.

#### Possible follow ups

There might be some improvements which I'll point to in further comments / questions in the code directly. In case you like this feature I can extend it to `R.hasIn()` as well. (For `R.hasPath()` I am not sure whether my TypeScript skills are sufficient. Thus I would like to try this in a new PR).

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ramda ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ramda@latest/))
was missing the following properties:
1. sequence
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 348858a6fa119599d8a8de88b4c2dfd37b2a5828
</p>

File: types/ramda/test/has-tests.ts [hunk 1]
```diff
@@ -1,13 +1,51 @@
...
 
 () => {
-  const hasName = R.has('name');
-  const a1: boolean = hasName({ name: 'alice' }); // => true
-  const a2: boolean = hasName({ name: 'bob' }); // => true
-  const a3: boolean = hasName({}); // => false
+    const hasName = R.has('name');
+    const a1: boolean = hasName({ name: 'alice' }); // => true
+    const a2: boolean = hasName({ name: 'bob' }); // => true
+    const a3: boolean = hasName({}); // => false
```
Comment (kulla) [inline types/ramda/test/has-tests.ts:7]: I noticed that some of the files in the ramda package aren't formatted with the current version of prettier at https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/.prettierrc.json

To not having unrelated file changes I haven't used prettier for my contributions at `types/ramda/tools.d.ts` and `types/ramda/index.d.ts`. Would it be okay run prettier there or shall this be in another PR?

**Review by Nicholaiii — APPROVED** [review]
Looks like a solid improvement to me.

File: types/ramda/tools.d.ts [hunk 2]
```diff
@@ -311,6 +311,21 @@ export type ObjPred<T = unknown> = (value: any, key: unknown extends T ? string
...
+ * ```
+ * // $ExpectType { foo: unknown } | { bar: unknown }
+ * type Foo = ObjectHavingSome<"foo" | "bar">
+ * ```
+ */
+// Implementation taken from
+// https://github.com/piotrwitek/utility-types/blob/df2502ef504c4ba8bd9de81a45baef112b7921d0/src/mapped-types.ts#L351-L362
+export type ObjectHavingSome<Key extends string> = A.Clean<{
+    [K in Key]: { [P in K]: unknown }
+}[Key]>;
```
Comment (kulla) [inline types/ramda/tools.d.ts:327]: I am quite new to advanced type declarations in typescript. Thus there might be a better implementation with types defined at https://github.com/millsp/ts-toolbelt I would be happy when you can point me to the appropriate utility function I can use here.

File: types/ramda/index.d.ts [hunk 3]
```diff
@@ -697,10 +698,10 @@ export function gte(a: number): (b: number) => boolean;
...
  * Returns whether or not an object has an own property with the specified name.
  */
-export function has<T>(__: Placeholder, obj: T): (s: string) => boolean;
-export function has<T>(__: Placeholder): (obj: T, s: string) => boolean;
-export function has<T>(s: string, obj: T): boolean;
-export function has(s: string): <T>(obj: T) => boolean;
+export function has(__: Placeholder, obj: unknown): (s: string) => boolean;
+export function has<P extends string>(__: Placeholder): (obj: unknown, s: P) => obj is ObjectHavingSome<P>;
+export function has<P extends string>(s: P, obj: unknown): obj is ObjectHavingSome<P>;
+export function has<P extends string>(s: P): (obj: unknown) => obj is ObjectHavingSome<P>;
```
Comment (kulla) [inline types/ramda/index.d.ts:704]: I have removed the `<T>` type variable here since I always got an error in the type predicate when it remained in the code. I also do not know which purpose this type variable has in the first place. However this might introduce a breaking change when some has used something like `const checkFunc = R.has<object>("name")` so that checkFunc has the signature `object => boolean`.

Comment (kulla) [reply types/ramda/index.d.ts:704]: Okay, I just recognized that `unknown => boolean` is compatible to any function `T => boolean` so that this should not be a problem.

**Comment by kulla** [comment]
> Looks like a solid improvement to me.

@Nicholaiii What do you think about https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50331#discussion_r549889029 The only breaking change I see is when somebody has used the type variable `T` explicitly: `R.has<object>("name", ...)` With the new implementation this would break since I remove the type variable `T` and introduce a new one for the key variable. However I do not see why anyone should have done this...

File: types/ramda/test/has-tests.ts (hunk 1 above)
Comment (Nicholaiii) [reply types/ramda/test/has-tests.ts:7]: > I noticed that some of the files in the ramda package aren't formatted with the current version of prettier at https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/.prettierrc.json
> 
> To not having unrelated file changes I haven't used prettier for my contributions at `types/ramda/tools.d.ts` and `types/ramda/index.d.ts`. Would it be okay run prettier there or shall this be in another PR?

If the changes are unrelated to this issue, you should definitely make a new PR if you can! That would be great too
